### PR TITLE
Tensor optimizations and rectangular matrices

### PR DESCRIPTION
**Pull request by pseewald**
These crucial optimizations make the tensor implementation ready for applications in CP2K. Most parts of cubic scaling [RPA code](https://github.com/cp2k/cp2k/blob/master/src/rpa_im_time.F) will be replaced by tensor operations.

### CODE REVIEW

**Comment by alazzaro** [comment]
Please wait before merging. This will be release 2.0

**Comment by alazzaro** [comment]
@pseewald the PR looks fine to me (only a single comment).
I see only a "design" problem: 
From what I understand all *rectangular* classes are only used within the tensors part. In principle, all those modules are not part of the dbcsr_api, so the users cannot use them. Would it make sense to add them to the dbcs_api and add an example on how to use them? Otherwise I propose to move the rectangular directory inside the tensors one (this is a pity, it would be good to make them public and add an example :) )

**Comment by pseewald** [comment]
@alazzaro in fact there is an example and I can commit it. Then I can also add rectangular types and routines to the API. Since the tensor is now part of DBCSR, would you say we can add tensors to DBCSR API as well (instead of having its own API)? Or do we plan for distinct APIs (i.e. another API for linear algebra...)

File: src/ops/dbcsr_operations.F [hunk 1]
```diff
@@ -2753,6 +2754,18 @@ FUNCTION dbcsr_get_occupation(matrix) RESULT(occupation)
...
    END FUNCTION dbcsr_get_occupation
 
+! **************************************************************************************************
+!> \brief Clear a matrix (remove all blocks)
+!> \param matrix ...
+! **************************************************************************************************
+   SUBROUTINE dbcsr_clear(matrix)
+      TYPE(dbcsr_type), INTENT(INOUT)                       :: matrix
+
+      CALL dbcsr_set(matrix, 0.0_dp)
```
[the remark this answers is not in the document]
Comment (pseewald) [reply src/ops/dbcsr_operations.F]: Good suggestion, will implement it like this

**Comment by pseewald** [comment]
I think it should be properly tested and documented for what kind of matrices the `rectangular` type is needed. To the DBCSR user it looks like a confusing design decision to have a separate type for rectangular matrices and people may think that DBCSR can not deal with rectangular matrices (which is not true of course). This new `rectangular` type is an optimization for very skinny matrices that are very sparse (in cubic-scaling RPA the largest matrices I tested have block dimensions ~ 10^8 x 3*10^3) and I think only for these extreme cases this module is really needed. I should probably commit a standalone test that can also serve as a benchmark for various matrix sizes such that the benefits of the new type can actually be tested. Also I think I should change naming - it seems that in the literature people use the term _tall-and-skinny_ matrices - so we could rename `rectangular` to `tas` and rename `dbcsr_r_*` into `dbcsr_tas_*`?

**Comment by pseewald** [comment]
> Well, currently the tensors part is using the dbcsr_api, so to have a full migration yo have to change all tensors modules to use directly the dbcsr modules. 

This is not entirely true - only routines that convert between tensors and DBCSR matrices uses dbcsr_api, the tensors itself are using `dbcsr_r_type` which in turn is using `dbcsr_type`. I'm not sure if this is good design or whether I should base rectangular matrices on `dbcsr_api`. As it is now and for consistency with the current API design I would have to make another wrapper around `dbcsr_r_type` in order to make its component (and components of `dbcsr_type`) private, is that correct? So it would actually be much easier to use `dbcsr_api` for rectangular matrices...

**Comment by alazzaro** [comment]
> I think it should be properly tested and documented for what kind of matrices the `rectangular` type is needed. To the DBCSR user it looks like a confusing design decision to have a separate type for rectangular matrices and people may think that DBCSR can not deal with rectangular matrices (which is not true of course). 

I agree that a new type specific for rectangular 'tas' matrices can confuse people. For this reason, I think it is better to keep them out from the dbcsr_api (changed my mind, sorry). Therefore, please discard my original request, no need to put them in the API and no need to have examples.

Now that v2.0.0 development is officially started, I would propose to merge this PR and then we can think about how to refactor the code. I think the rectangular directory (renamed to 'tas', that's good) should go inside tensors directory and everything should be consistent with the dbcsr_api (which will be refactored too). Anyway, this is something we can do after the merge.

> This new `rectangular` type is an optimization for very skinny matrices that are very sparse (in cubic-scaling RPA the largest matrices I tested have block dimensions ~ 10^8 x 3*10^3) and I think only for these extreme cases this module is really needed. I should probably commit a standalone test that can also serve as a benchmark for various matrix sizes such that the benefits of the new type can actually be tested. Also I think I should change naming - it seems that in the literature people use the term _tall-and-skinny_ matrices - so we could rename `rectangular` to `tas` and rename `dbcsr_r_*` into `dbcsr_tas_*`?

Concerning the tests for the performance, this is should be part of the current "framework" `dbcsr_performance_driver.F`, so that we can have a fast comparison between "standard" and "custom" algorithms. Let's do that after the merge too.

Concerning the names, well, I don't have any strong option on that...

**Comment by alazzaro** [comment]
> > Well, currently the tensors part is using the dbcsr_api, so to have a full migration yo have to change all tensors modules to use directly the dbcsr modules.
> 
> This is not entirely true - only routines that convert between tensors and DBCSR matrices uses dbcsr_api, the tensors itself are using `dbcsr_r_type` which in turn is using `dbcsr_type` of `dbcsr_types`. I'm not sure if this is good design or whether I should base rectangular matrices on `dbcsr_api`. As it is now and for consistency with the current API design I would have to make another wrapper around `dbcsr_r_type` in order to make its component (and components of `dbcsr_type`) private, is that correct? So it would actually be much easier to use `dbcsr_api` for rectangular matrices...

Let's discuss this in a different PR (rafactoring)... Now that we are in the v2 we can think to change the API...

**Comment by alazzaro** [comment]
In conclusion, besides my comment on the dbcsr_clear, I think the only missing part is an example (or more examples?) on how to use the tensors. After that, this PR is ready to merge. DBCSR v2 will be the target. Well done work @pseewald !

**Comment by pseewald** [comment]
I don't have a meaningful tensor example ready (it's a bit less straightforward than writing examples for matrix multiplication) and I'd like to commit this in a separate pull request later. Is there anything else?

**Comment by alazzaro** [comment]
Ok, please merge. Examples can added later (you can also change the readme.md to reflect that we do tensors too)

**Pull request merged**